### DISCUSSION TITLE
Using Flask and Flask-Restful with Lambda Power Tools

### DISCUSSION THREAD

**Discussion by BwL1289** [aws-powertools/powertools-lambda-python#1771] [Q&A]
Hello! Apologies if this has been covered elsewhere, but I've looked and haven't found anything explicitly addressing it, and I can't find anything that explicitly says it can't be done.

Currently using Flask and Flask-Restful, and am slowly incorporating lambda power tools into the stack. My question is: is it possible to use lambda power tools with Flask (with or without Flask-Restful)?

I understand it's not a full replacement for Flask, but i'd love to incorporate the convenience of powertools into our stack to get to "well architected" even if I needed to rip out Flask-Restful. 

I know there's some functionality that simply won't work without the `event handler` (i.e. raising HTTP errors, etc), but I would hope that some functionality like the below would be available:

```
class ExampleAPI(Resource):
    @metrics.log_metrics(capture_cold_start_metric=True, raise_on_empty_metrics=True)
    @tracer.capture_lambda_handler
    @logger.inject_lambda_context(log_event=True, correlation_id_path=correlation_paths.API_GATEWAY_REST)
    def post(self):
        secret = parameters.get_secret(...)
        ...
```
Looking at the source code, it's **completely obvious why this is happening**, but wanted to open a discussion to see if anyone has found workarounds without migrating an entire codebase at once.

Consistent error I'm seeing:
`TypeError: decorate() missing 1 required positional argument: 'context'`

Cheers!

**Comment by heitorlessa** [comment]
Hey @BwL1289 - would you be able to share a complete sample so we can tinker it through?

From the top of my head, you could:

* **Extend your method signature to accept a default Lambda Context**. Since Lambda Handler is always the entrypoint, you could keep a reference in Flask G (globals) for every request, and use dependency injection. This will make Logger and Metrics decorators work as expected.
* **Use `tracer.capture_method` instead of `tracer.capture_lambda_handler`**. [The former is specifically for any Python synchronous, asynchronous and generator functions](https://awslabs.github.io/aws-lambda-powertools-python/2.4.0/core/tracer/#synchronous-functions) (hope to add a `capture_method_async` in the future).

It's been a while since I last used Flask-Restful so a working (deployable) example would be great to experiment a pattern we could document for everyone.

Thanks a lot for kicking this discussion!

PS: We might make better progress in January due to winter holidays for all maintainers (I'll put up a notice this week).

**Reply by BwL1289** [reply]
@heitorlessa Thanks so much, and yessir! I will provide a complete example and post it here as soon as possible. I will also try out some of your suggestions above.

**Reply by heitorlessa** [reply]
Awesome. Look forward to hearing from you.

Cc @leandrodamascena who's back from the maintainers side

**Reply by leandrodamascena** [reply]
Hey all, I'm back in the game! @BwL1289 pls let us know when you're ready to send us the full example :)

**Reply by BwL1289** [reply]
Will do! Appreciate your patience during the holidays.

**Comment by BwL1289** [comment]
Hey @leandrodamascena! I had to temporarily move to a different (albeit related) project, so apologies for the delay on getting you the full example. For the future, please let me know if it's OK to send you the example privately due to privacy concerns. Of course, let's post the solution here so everyone can learn from it!

A bit off-topic, but related, to this discussion — we're using powertools in another project we're working on, but we'd like to use it on methods in a class. This class does not use Flask-Restful (just a regular python class). I haven't found any examples and can't find a way to avoid the mismatch between what the decorators expect as a first arg (`lambda_handler`) and `self`. Example:

```
class ThumbprintHandler:
    @tracer.capture_method
    def create_thumbprint(self) -> list[str]:
         ....
```

PyLance reports: `Parameter 1: type "Dict[Unknown, Unknown]" cannot be assigned to type "Self@ThumbprintHandler"`

Again, it's blatantly obvious why it's happening, but can't find a workaround. Apologies in advance if it's obvious how to do so and I'm missing something, but sincerely appreciate the help!

**Comment by leandrodamascena** [comment]
Hi @BwL1289, sorry for the late reply! We had internal meetings and stuff and things were really busy here.

Yes, you can send me the project directly through Discord or another option that you find most suitable.

Related to Pylance, I couldn't simulate here. I created a class and method exactly like you and Pylance didn't complain about anything. Can you send me a screenshot or full example to verify this? `capture_method` is a function overload and you can use it to decorate your functions outside the lambda, even inside a class.

Thank you.

**Reply by heitorlessa** [reply]
following up @BwL1289 - did you manage to get this solved? can we still help?

**Comment by leandrodamascena** [comment]
I'm closing this discussion as resolved. Please reopen if you need.
